base/subarray.jl

```diff
@@ -347,71 +211,37 @@ first(::Colon) = 1
 isempty(::Colon) = false
 in(::Int, ::Colon) = true
 
-## Strides
-@generated function strides{T,N,P,I}(V::SubArray{T,N,P,I})
-    Ip = I.parameters
-    all(x->x<:Union{RangeIndex,Colon}, Ip) || throw(ArgumentError("strides valid only for RangeIndex indexing"))
-    strideexprs = Array(Any, N+1)
-    strideexprs[1] = 1
-    i = 1
-    Vdim = 1
-    for i = 1:length(Ip)
-        if Ip[i] != Int
-            strideexprs[Vdim+1] = copy(strideexprs[Vdim])
-            strideexprs[Vdim] = :(step(V.indexes[$i])*$(strideexprs[Vdim]))
-            Vdim += 1
-        end
-        strideexprs[Vdim] = :(size(V.parent, $i) * $(strideexprs[Vdim]))
-    end
-    :(tuple($(strideexprs[1:N]...)))
-end
+# Strides are the distance between adjacent elements in a given dimension,
+# so they are well-defined even for non-linear memory layouts
+strides{T,N,P,I}(V::SubArray{T,N,P,I}) = substrides(1, V.parent, 1, V.indexes...)
+substrides(s, parent, dim) = ()
+substrides(s, parent, dim, i::DroppedScalar, I...) = (substrides(s*size(parent, dim), parent, dim+1, I...)...)
+substrides(s, parent, dim, i::NoSlice, I...) = (s, substrides(s*size(parent, dim), parent, dim+1, I...)...)
+substrides(s, parent, dim, i::Union{Range, Colon}, I...) = (s*step(i), substrides(s*size(parent, dim), parent, dim+1, I...)...)
+substrides(s, parent, dim, i, I...) = throw(ArgumentError("strides is invalid for SubArrays with indices of type $(typeof(i))"))
 
 stride(V::SubArray, d::Integer) = d <= ndims(V) ? strides(V)[d] : strides(V)[end] * size(V)[end]
 
-function stride1expr(Atype::Type, Itypes, Aexpr, Isym, LD, Jsym=Isym, Iindex_lin=0, Jindex_lin=0)
-    if LD == 0
-        return 0
-    end
+function stride1expr(Itypes, Aexpr, Isym)
```

Comment:
Ah, yes, I had some un-committed work that didn't make it into this patch that does so.